src/platform/Darwin/DnssdImpl.h

```diff
@@ -33,29 +32,70 @@ enum class ContextType
     Register,
     Browse,
     Resolve,
-    GetAddrInfo,
 };
 
 struct GenericContext
 {
     ContextType type;
     void * context;
     DNSServiceRef serviceRef;
+
+    virtual ~GenericContext() {}
+
+    CHIP_ERROR Finalize(DNSServiceErrorType err = kDNSServiceErr_NoError);
+    virtual void DispatchFailure(DNSServiceErrorType err) = 0;
+    virtual void DispatchSuccess()                        = 0;
+};
+
+class MdnsContexts
+{
+public:
+    MdnsContexts(const MdnsContexts &) = delete;
+    MdnsContexts & operator=(const MdnsContexts &) = delete;
+    ~MdnsContexts();
+    static MdnsContexts & GetInstance() { return sInstance; }
+
+    CHIP_ERROR Add(GenericContext * context, DNSServiceRef sdRef);
+    CHIP_ERROR Remove(GenericContext * context);
+    CHIP_ERROR RemoveAllOfType(ContextType type);
+    CHIP_ERROR Has(GenericContext * context);
+
+    /**
+     * @brief
+     *   Returns a pointer to a context that has previously been registered
+     *   with a given type.
```

Comment:
```suggestion
     *   Returns a pointer to a RegisterContext that has previously been registered
     *   with a given type.
```
And maybe we should actually have the outparam be `RegisterContext **`?